Minimum size for shape recognizer now customizable

hyperupcall opened the pull request:
When "shape recognizer" mode is enabled, and when a "pen stroke happens", the new code does the following:
  - take the minimum and maximum values for both the X and Y axis
  - if the bigger X value minus the smaller X value is less than the "minimum size" (and if this holds true for the Y values too), then it does not draw the shape

Note that some of the changes in the diff are due to `clang-format`

Closes #2531

Validation:
- Manual testing

bhennion commented on ui/settings.glade:
```diff
@@ -1158,7 +1170,7 @@ If the file was not yet saved you can find it in ~/.cache/xournalpp/autosaves&lt
                                             <property name="visible">True</property>
                                             <property name="can-focus">False</property>
                                             <property name="tooltip-text" translatable="yes">Gaussian parameter: the higher, the more stabilization and the more lag between your cursor and the painting tip.</property>
-                                            <property name="label" translatable="no">σ</property>
```
Did you change this on purpose?

hyperupcall replied:
Oops! I must have accidentally pressed something within the Glade UI - I'll check the diff next time so this doesn't sneak in again...

bhennion commented on src/core/control/shaperecognizer/ShapeRecognizer.cpp:
```diff
@@ -242,13 +244,43 @@ void ShapeRecognizer::optimizePolygonal(const Point* pt, int nsides, int* breaks
+        return false;
+    }
+
+    double Xmax = std::numeric_limits<double>::min();
+    double Xmin = std::numeric_limits<double>::max();
+
+    double Ymax = std::numeric_limits<double>::min();
+    double Ymin = std::numeric_limits<double>::max();
+
+    for (auto& point: stroke->getPointVector()) {
+        if (point.x > Xmax)
+            Xmax = point.x;
+        if (point.x < Xmin)
+            Xmin = point.x;
+
+        if (point.y > Ymax)
+            Ymax = point.y;
+        if (point.y < Ymin)
+            Ymin = point.y;
+    }
```
You can retrieve this size by calling `stroke->getSnappedBounds()` The calculation is the same (thus removing code duplication), and the result is cached for later use (in Element::snappedBounds).

hyperupcall replied:
:+1:

bhennion reviewed (commented):
Just a couple of tiny remarks here

hyperupcall commented:
Alright, fixed those two issues! I'll also add that I changed the default value from `30` to `40`, since it's larger than the height of the page lines of the default background (rather than equal to).

bhennion commented:
The code LGTM. I'm a bit scared the change in behaviour might surprise regular users of the stroke recognizer.
@xournalpp/core what do you think?

hyperupcall commented:
Just an info on the recent activity - it rebases over `master` since there were conflicts

Technius commented:
> I'm a bit scared the change in behaviour might surprise regular users of the stroke recognizer.

If the default value is set to approximately the size of strokes ignored by the old behavior ("size" of three points), it should be fine. Not sure what "size 40" corresponds to.

Technius commented on src/core/control/shaperecognizer/ShapeRecognizer.h:
```diff
@@ -37,6 +37,8 @@ class ShapeRecognizer {
 
     int findPolygonal(const Point* pt, int start, int end, int nsides, int* breaks, Inertia* ss);
 
+    bool isStrokeLargeEnough(Stroke* stroke, double strokeMinSize);
```
Needs documentation. Also, this can be a static function.

Technius commented on src/core/control/shaperecognizer/ShapeRecognizer.cpp:
```diff
@@ -242,13 +244,26 @@ void ShapeRecognizer::optimizePolygonal(const Point* pt, int nsides, int* breaks
+    }
+
+    auto rect = stroke->getSnappedBounds();
+    if (rect.width < strokeMinSize && rect.height < strokeMinSize) {
+        return false;
+    }
+
+    return true;
```
nit: can be simplified to
```suggestion
    return rect.width >= strokeMinSize || rect.height >= strokeMinSize;
```

Technius replied:
It looks like this only checks the axis-aligned bounding box. If someone sets a large minimum and draws diagonal lines, will there be a noticeable or awkward difference compared to drawing horizontal or vertical lines?

hyperupcall replied:
Hm yeah there is no difference between drawing diagonal lines and completely vertical or horizontal lines because the condition only activates when both axis are smaller than the `strokeMinSize`. I don't think it'll be unnatural unless I'm not understanding your question correctly

hyperupcall replied:
About the nit - I'm not sure if the simplification helps readability. It continues the pattern of the previous if-statement and the `<` makes it directly obvious that the width and lengths must both be within the bounding box - the suggested changes make it a bit less clear and there is a little more thinking involved. And it might be a bit of a stretch, but it might be more error prone when adding a conditional in the future, as the conditional would have to be DeMorganized again. But if you wish, shall I still make those changes?

rolandlo replied:
> Hm yeah there is no difference between drawing diagonal lines and completely vertical or horizontal lines because the condition only activates when both axis are smaller than the `strokeMinSize`. I don't think it'll be unnatural unless I'm not understanding your question correctly

It might be more natural to use:
```return std::hypot(rect.width, rect.height) >= strokeMinSize```
(length of the diagonal of the rectangle)

hyperupcall commented:
I switched to `std::hypot` as suggested (to me there doesn't seem to be a huge difference, but the hypot might be more accurate when drawing lines, compared to circles and squares). i also cleaned up the `return` :+1:

Technius reviewed (approved):
LGTM. Last point of concern might be the min size -- a size of `0` corresponds to the old default (which people did not find helpful based on the issue linked to this PR) -- but a good default is something that we'll have to determine through trial and error.

Technius commented:
Merging in 24 hours if there are no objections.

bhennion commented:
Thanks for the great work!
